### DISCUSSION TITLE
Does /tools/docker/debian support networking using the network_relay_url?

### DISCUSSION THREAD

**Discussion by nospaceleftondevice** [copy/v86#532] [Q&A]
I built the debian image via docker, but I cant seem to get networking working.  I see the networking.sh script which enables the interface with a static ip, but I see nothing hitting the relay.  I have not tried to use DHCP, has anyone got this working?

**Comment by copy** [comment]
First, check that the ne2k module is loaded: `lsmod` should include `ne2k_pci`. Then check that the card is detected: `ifconfig -a` should print `eth0` or similar. Then run `dhcpcd -w4 eth0`.

**Reply by BradAtVivSoft** [reply]
That most certainly did the trick, thank you.

**Reply by ZXMushroom63** [reply]
I wrote a shell script to automate it:
```
#!/bin/bash

STR=$(lsmod)
SUB='ne2k_pci'

if [[ "$STR" =~ .*"$SUB".* ]]; then
   NETID=$(head -n 1 <<< $(ifconfig -a | sed 's/[ \t].*//;/^\(lo\|\)$/d' | sed 's/\://'));
   echo "Detected ne2k, using "$NETID" as dhcpcd target."
   sudo dhcpcd -w4 $NETID
else
   echo "This system is not running in v86."
fi

```

**Comment by SugarRayLua** [comment]
@nospaceleftondevice, would you mind sharing the steps you used to get a Debian image using the /tools/docker/Debian and uploading it to v86? I'm a novice linux user and git cloned the v86 repository on my Mac terminal, build the Debian-full docker file and then went looking for the Debian image to upload to the v86 when it allows you to "choose an image" on startup. I don't see any ".iso" in the Debian-full docker container after it is built that I could use to upload to v86.  Thanks!

**Reply by catplanet007** [reply]
`make run` then open http://127.0.0.1:8000/examples/debian.html

**Comment by owenthereal** [comment]
I couldn't get networking to work with the [debian examples](https://github.com/copy/v86/tree/master/tools/docker/debian). Could anyone help?

What I did:


* I made the following changes:

```diff
diff --git a/examples/debian.html b/examples/debian.html
index 5ee2dabf..1b759168 100644
--- a/examples/debian.html
+++ b/examples/debian.html
@@ -12,6 +12,9 @@ window.onload = function()
         memory_size: 512 * 1024 * 1024,
         vga_memory_size: 8 * 1024 * 1024,
         screen_container: document.getElementById("screen_container"),
+        network_relay_url: "ws://localhost:8080",
+        preserve_mac_from_state_image: true,
+        mac_address_translation: true,
         initial_state: { url: "../images/debian-state-base.bin" },
         filesystem: { baseurl: "../images/debian-9p-rootfs-flat/" },
         autostart: true,

diff --git a/tools/docker/debian/build-state.js b/tools/docker/debian/build-state.js
index f90fabbf..6f7958a1 100755
--- a/tools/docker/debian/build-state.js
+++ b/tools/docker/debian/build-state.js
@@ -26,9 +26,13 @@ var emulator = new V86({
     autostart: true,
     memory_size: 512 * 1024 * 1024,
     vga_memory_size: 8 * 1024 * 1024,
-    network_relay_url: "<UNUSED>",
     bzimage_initrd_from_filesystem: true,
     cmdline: "rw init=/bin/systemd root=host9p console=ttyS0 spectre_v2=off pti=off",
+    network_relay_url: "ws://localhost:8080",
+    preserve_mac_from_state_image: true,
+    mac_address_translation: true,
     filesystem: {
         basefs: {
             url: path.join(V86_ROOT, "/images/debian-base-fs.json"),
```
* I followed the [README](https://github.com/copy/v86/blob/master/tools/docker/debian/Readme.md) to build the image and export the state
* I booted the ws proxy following https://github.com/copy/v86/blob/master/docs/networking.md
* I ran `sh networking.sh` after the booting the image on the browser
![Screenshot 2024-01-31 at 3 12 44 PM](https://github.com/copy/v86/assets/169064/9e89e3ff-9127-4fd1-ada4-96d65159053b)
* I see that the ws proxy has a client connected
* `curl google.com` returns `curl: (6) Could not resolve host: google.com`

**Reply by owenthereal** [reply]
I also tried the `dhcpcd -w4 enp0s5` but still not working:

![Screenshot 2024-01-31 at 5 18 21 PM](https://github.com/copy/v86/assets/169064/4a3613aa-6cbc-44ff-8ec3-c25a0991bfca)

**Reply by copy** [reply]
That looks like everything is working fine. Try `dig @8.8.8.8 google.com` or `ping 8.8.8.8`

**Reply by owenthereal** [reply]
`dig @8.8.8.8 google.com` works but `curl google.com` hangs and eventually times out

![Screenshot 2024-01-31 at 6 43 07 PM](https://github.com/copy/v86/assets/169064/a91b680d-d45e-4c25-a7e1-a0fac2bbfa6d)

**Reply by owenthereal** [reply]
![Screenshot 2024-01-31 at 7 03 53 PM](https://github.com/copy/v86/assets/169064/25d89b59-85b4-4076-b394-3a1f7b742517)

**Reply by owenthereal** [reply]
I finally got it to work. The issue is related to the local relay server. I switched to the demo relay server (`wss://relay.widgetry.org/`), and I could connect 😃 . However, the internet speed is insanely slow:

![Screenshot 2024-02-01 at 10 55 27 AM](https://github.com/copy/v86/assets/169064/887097b9-6af7-494a-9d15-96b69aacc19f)

I only got `40kB/s` and am curious if there is any way to improve it. Has anyone had any success in running the relay server locally and getting the internet to work?

**Reply by BelleNottelling** [reply]
The relay is both very slow and also rate limited. You're unlikely to get much better than 40KB/s even if you run a local version with the rate limit removed.

The best speeds I've achieved were with the [node relay](https://github.com/BelleNottelling/node-relay) (my fork primarily changes the DNS from google to quad9), but even then I think the peak I saw was only a few MB/s. I've thought about hosting an instance of it, but the original version doesn't have a provided license and no clarification was provided when I asked about it.

**Reply by owenthereal** [reply]
Thanks for the tips @BelleNottelling! I wonder whether v86 can access the internet without a relay server (https://github.com/copy/v86/issues/198). Removing the need for a relay server would make v86 more usable and portable.

**Reply by BelleNottelling** [reply]
I'd assume probably not in the near future, at least not officially. Maybe some day some new tech will make it more feasible, but at the moment there's no clean solution AFAIK.
